### ISSUE TITLE
model.py — numerical stability, log-prob naming, and validation fixes

### ISSUE THREAD

**Issue by stepanyanhayk** [McDermottHealthAI/MedRAP#58]
> **Depends on**: #51 (EncoderOutput.attention_mask), #52 (QueryProjector interface), #55 (fusion compatibility guard), #56 (pooling mask threading)

## ML fixes

**`_marginal_class_probabilities` — numerically unstable inference path**
Direct probability multiplication underflows when K is large or softmax scores are sharply peaked. `clamp_min(1e-8)` silently masks the underflow with `log(1e-8) = -18.4`. The training losses correctly use `logsumexp` — inference should too:
```python
# current (unstable)
p_ret = softmax(doc_scores)
p_pred = softmax(per_doc_logits)
logits = (p_ret.unsqueeze(-1) * p_pred).sum(dim=1).clamp_min(1e-8).log()

# correct
log_p_ret = log_softmax(doc_scores, dim=-1)
log_p_pred = log_softmax(per_doc_logits, dim=-1)
logits = logsumexp(log_p_ret.unsqueeze(-1) + log_p_pred, dim=1)
```

**`logits` from the marginalized path are log-probabilities, not logits**
The marginalized output is in `(-∞, 0]`, not pre-sigmoid. Pairing `marginalized_retrieval=True` with `MultiTaskBCELoss` applies `sigmoid(log(p))` instead of `p` — silently wrong predictions and gradients with no error raised. Document explicitly in the class docstring that `MultiTaskBCELoss` is incompatible with `marginalized_retrieval=True` and that `MultiTaskBCEMarginalizedLoss` must be used instead.

**`self.pooling` silently bypassed in the marginalized path**
When `marginalized_retrieval=True`, `self.pooling` is never called — it is a registered submodule with parameters that have no effect. Document this explicitly in the `RetrievalAugmentedModel` docstring.

**No guard for incompatible fusion with `marginalized_retrieval`**
Only `fused_state.ndim == 3` is checked. `PassthroughFusion` and `CrossAttentionFusion` both return `(B, S_ehr, D)` which passes — the model silently marginalizes over EHR sequence positions instead of retrieved documents. Add an explicit check (see #55).

## SWE fixes

**`isinstance(self.head, LinearHead)` check should move to `__init__`**
This validation only depends on constructor arguments. Moving it to `__init__` catches misconfiguration at construction time rather than at first forward pass.

**`doc_scores` shape validated after the expensive computation**
The shape check at line 298 runs after `differentiable_retrieval_scores()` has already executed. Pre-validate that `R == 1` before calling the function to fail fast.

**Redundant re-imports in the second doctest block**
Lines 117–126 re-import `torch`, `MEDSTorchBatch`, `MEDSCodeEncoder`, `ReplaceFusion`, etc. — all already in scope from the first block. Remove them.

**`query_projector` and `pooling` call sites drop `attention_mask`**
```python
query_out = self.query_projector(encoder_out.patient_state)  # drops mask
pooled = self.pooling(fusion_out.fused_state)                # drops mask
```
Fix once #51, #52, and #56 are resolved.

## Tests

**`_marginal_class_probabilities` — entirely untested**
Add a doctest verifying that K=1 reduces to standard `softmax` (matching the analogous K=1 tests in both marginalized loss functions).

**Add non-marginalized gradient flow doctest**
The marginalized path has a `loss.backward()` + grad check (lines 160–165). Add an equivalent for the standard pooling path to verify end-to-end differentiability.
